// @neuralegion/class-sanitizer/dist/sanitazion/SanitizeTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SanitizeTypes = void 0;
/**
 * Sanitization types
 */
var SanitizeTypes;
(function (SanitizeTypes) {
    SanitizeTypes["BLACKLIST"] = "blacklist";
    SanitizeTypes["ESCAPE"] = "escape";
    SanitizeTypes["SECURE"] = "secure";
    SanitizeTypes["LTRIM"] = "ltrim";
    SanitizeTypes["NORMALIZE_EMAIL"] = "normalizeEmail";
    SanitizeTypes["RTRIM"] = "rtrim";
    SanitizeTypes["STRIP_LOW"] = "stripLow";
    SanitizeTypes["TO_BOOLEAN"] = "toBoolean";
    SanitizeTypes["TO_DATE"] = "toDate";
    SanitizeTypes["TO_FLOAT"] = "toFloat";
    SanitizeTypes["TO_INT"] = "toInt";
    SanitizeTypes["TO_STRING"] = "toString";
    SanitizeTypes["TRIM"] = "trim";
    SanitizeTypes["WHITELIST"] = "whitelist";
    SanitizeTypes["TO_LOWER_CASE"] = "toLowerCase";
    SanitizeTypes["TO_UPPER_CASE"] = "toUpperCase";
    SanitizeTypes["CUSTOM_SANITIZATION"] = "customSanitization";
    SanitizeTypes["NESTED"] = "nestedSanitization";
})(SanitizeTypes = exports.SanitizeTypes || (exports.SanitizeTypes = {}));
// eslint-disable-next-line @typescript-eslint/no-namespace
(function (SanitizeTypes) {
    function isValid(type) {
        return Object.values(SanitizeTypes).some(value => value === type);
    }
    SanitizeTypes.isValid = isValid;
})(SanitizeTypes = exports.SanitizeTypes || (exports.SanitizeTypes = {}));
